### ISSUE TITLE
Guidance on evaluating shard count/size, WebDataset length

### ISSUE THREAD

**Issue by tottenjordan** [webdataset/webdataset#75]
Trying to optimize TPU pod slice solution discussed in https://github.com/webdataset/webdataset/issues/47

Assuming scenario where you need to replicate training with epochs i.e., declare length of dataset, distribute equal samples/batches to nodes...

**(1) What is length actually doing to the Webdataset dataset or data loading? And how should it be calculated**

(a) My original understanding was this was the total number of samples for a particular worker in an epoch, calculated like this: 
> ```length={SHARD_SAMPLES} * {SHARDS_PER_WORKER}```

(b) Then in [older docs](https://github.com/tmbdev-archive/webdataset-examples/blob/master/main-wds.py#L330) I saw `length=num_batches`:
> ```num_batches = dataset_size // (batch_size * world_size)```

- Is the difference between these two recommendations whether you use `.batched(batch_size)` on the `WebDataset`? e.g., if using `.batched()` then calculate with (b)? 

**(2) Would calculation to `length` change if using both:**
-  `.batched(batch_size, partial=True)` on dataset 
- `torch.utils.data.DataLoader(dataset, batch_size=None, shuffle=False, drop_last=False, num_workers=num_workers)` 

**Recently realized I've actually been calculating `length` by the sample as well as batching:**
  - `num_dataset_instances = xm.xrt_world_size() * FLAGS.num_workers` (256=32*8)
  - `epoch_size = trainsize // num_dataset_instances` (5000=1280000//256)
  - `.batched(batch_size, partial=True)`

**(3) Any guidance or more thought around deciding optimal number of shards and size of shards (samples and/or bytes)?**
 - e.g., for ImageNet
 - I suspect one constraint for shard count is that it needs to be divisible by `num_workers * world_size`
 - At some point, as the # of shards increases/approaches # of samples, advantages of sharding decrease
 - Lower # of shards = large files, could be impractical for large datasets 
 - If we do lowest shard count divisible by `num_workers * world_size`:
   -  we make minimum storage requests, get more Sequential, able to parallel read/process, and distribute to accelerator
   - But this means the shards are gong to be large (especially when working with +TBs) 
   - Is this decision also related to memory/network constraints?

**Comment by mynameisvinn** [comment]
(2) For a dataset of length N, calling batch with batchsize B returns a new iterable of length equal to N/B. If your `DataLoader`'s batch_size is non-zero, then it will return batches of batches. I believe that's why it is recommended batching in the `Dataset` rather than in a `Dataloader`.

**Comment by tmbdev** [comment]
As far as WebDataset is concerned, the length does not matter, anywhere. WebDataset just iterates through data sources until it reaches the end and then raises a StopIteration. You can set it to whatever your framework requires.

All the constraints on actual length and nominal length come from whatever framewrk you are using for training, and from the need to have the same number of batches across all nodes participating in DistributedDataParallel _while_ at the same time training exactly by epoch.

So, if you just want an easy way of training, do this:

```
dataset = WebDataset(...).repeat().decode(...).to_tuple(...).batched(1024)
loader = WebLoader(dataset, ...)

for i, batch in loader:
    if i%10000 == 0:
        run_validation_step()
    ... train on batch ...
```

No need to ever deal with lengths, partial batches, etc.  This works regardless of the number of training nodes you use. You can also use resampling rather than iteration:

```
dataset = WebDataset(..., shardlist=ResampledShards).repeat().decode(...).to_tuple(...).batched(1024)
```

If your framework requires epochs, you can do the following:

```
dataset = WebDataset(...).repeat().decode(...).to_tuple(...).batched(1024)
loader = WebLoader(dataset, ...).repeat(nbatches=10000)

for epoch in range(nepochs):
    for i, batch in loader:
        ... train on batch ...
    run_validation_step()
```

If you really must have absolutely, perfectly exact epochs, the default will work fine for single node training:

```
dataset = WebDataset(...).decode(...).to_tuple(...).batched(1024)
loader = WebLoader(dataset, ...)

for epoch in range(nepochs):
    for i, batch in loader:
        ... train on batch ...
    run_validation_step()
```

You can't do exact epoch training for multinode training using DistributedDataParallel by only changing the I/O pipeline. The closest that any I/O pipeline can get you is via padding and truncating. This does require you to specify the number of batches after which things should get truncated.

```
dataset = WebDataset(...).decode(...).to_tuple(...).batched(1024)
loader = WebLoader(dataset, ...).ddp_equalize(dataset_size // batch_size)

for epoch in range(nepochs):
    for i, batch in loader:
        ... train on batch ...
    run_validation_step()
```

So, in short, set the length to whatever you like.

I'm going to try to document this better and clean up some of the DDP-related options.

**Comment by HuangChiEn** [comment]
> As far as WebDataset is concerned, the length does not matter, anywhere. WebDataset just iterates through data sources until it reaches the end and then raises a StopIteration. You can set it to whatever your framework requires.
> 
> All the constraints on actual length and nominal length come from whatever framewrk you are using for training, and from the need to have the same number of batches across all nodes participating in DistributedDataParallel _while_ at the same time training exactly by epoch.
> 
> So, if you just want an easy way of training, do this:
> 
> ```
> dataset = WebDataset(...).repeat().decode(...).to_tuple(...).batched(1024)
> loader = WebLoader(dataset, ...)
> 
> for i, batch in loader:
>     if i%10000 == 0:
>         run_validation_step()
>     ... train on batch ...
> ```
> 
> No need to ever deal with lengths, partial batches, etc. This works regardless of the number of training nodes you use. You can also use resampling rather than iteration:
> 
> ```
> dataset = WebDataset(..., shardlist=ResampledShards).repeat().decode(...).to_tuple(...).batched(1024)
> ```
> 
> If your framework requires epochs, you can do the following:
> 
> ```
> dataset = WebDataset(...).repeat().decode(...).to_tuple(...).batched(1024)
> loader = WebLoader(dataset, ...).with_epoch(10000)
> 
> for epoch in range(nepochs):
>     for i, batch in loader:
>         ... train on batch ...
>     run_validation_step()
> ```
> 
> If you really must have absolutely, perfectly exact epochs, the default will work fine for single node training:
> 
> ```
> dataset = WebDataset(...).decode(...).to_tuple(...).batched(1024)
> loader = WebLoader(dataset, ...)
> 
> for epoch in range(nepochs):
>     for i, batch in loader:
>         ... train on batch ...
>     run_validation_step()
> ```
> 
> You can't do exact epoch training for multinode training using DistributedDataParallel by only changing the I/O pipeline. The closest that any I/O pipeline can get you is via padding and truncating. This does require you to specify the number of batches after which things should get truncated.
> 
> ```
> dataset = WebDataset(...).decode(...).to_tuple(...).batched(1024)
> loader = WebLoader(dataset, ...).ddp_equalize(dataset_size // batch_size)
> 
> for epoch in range(nepochs):
>     for i, batch in loader:
>         ... train on batch ...
>     run_validation_step()
> ```
> 
> So, in short, set the length to whatever you like.
> 
> I'm going to try to document this better and clean up some of the DDP-related options.

So, suppose I have 1 node 2 gpus for distributed training in DDP mode, the following example will automatically assign the non-overlapped data slice for the corresponding GPU ?
 And, I can set dataset_size as a random number, like 123456, and it will ignore this setup ?
( because we can not get the real dataset_size in webdataset, it's "iterator dataset".

```
dataset = WebDataset(...).decode(...).to_tuple(...).batched(1024)

# I can set dataset_size as a random number, like 123456 ? 
loader = WebLoader(dataset, ...).ddp_equalize(dataset_size // batch_size)

for epoch in range(nepochs):
    for i, batch in loader:
        ... train on batch ...
    run_validation_step()

```

**Comment by tmbdev** [comment]
You are not repeating your training data infinitely, so this won't work.

If you want exactly one permutation of the training data per epoch and you want this distributed equally across the GPUs, you should use the `wids` portion of the WebDataset library; it emulates an indexed dataset perfectly and works with the multinode-related code in the PyTorch DataLoader.

https://github.com/webdataset/webdataset/blob/main/examples/train-resnet50-multiray-wds.ipynb

If you don't care about one permutation of the training data per epoch (and there is no reason you should; it does not result in better models), you just use WebDataset(..., resampled=True). In that case, just use `with_epoch` of whatever you find convenient, or get rid of epochs altogether, since they serve no purpose anymore.

`ddp_equalize` was an attempt to work around PyTorch's limited support for distributed training with IterableDataset. It kind of works, but it is never quite exact and is not recommended anymore.

**Comment by HuangChiEn** [comment]
> You are not repeating your training data infinitely, so this won't work.
> 
> If you want exactly one permutation of the training data per epoch and you want this distributed equally across the GPUs, you should use the `wids` portion of the WebDataset library; it emulates an indexed dataset perfectly and works with the multinode-related code in the PyTorch DataLoader.
> 
> https://github.com/webdataset/webdataset/blob/main/examples/train-resnet50-multiray-wds.ipynb
> 
> If you don't care about one permutation of the training data per epoch (and there is no reason you should; it does not result in better models), you just use WebDataset(..., resampled=True). In that case, just use `with_epoch` of whatever you find convenient, or get rid of epochs altogether, since they serve no purpose anymore.
> 
> `ddp_equalize` was an attempt to work around PyTorch's limited support for distributed training with IterableDataset. It kind of works, but it is never quite exact and is not recommended anymore.

I think we'll take this good advice. Currently, we apply `resampled=True` and `with_epoch`. thanks for your reply.
